src/dill/anySwap/AnyCallClient.sol

```diff
@@ -0,0 +1,532 @@
+
+    event LogSwapout(
+        address indexed token,
+        address indexed sender,
+        address indexed receiver,
+        uint256 amount,
+        uint256 toChainId,
+        int256[] weights
+    );
+    event LogSwapin(
+        address indexed token,
+        address indexed sender,
+        address indexed receiver,
+        uint256 amount,
+        uint256 fromChainId,
+        int256[] weights
+    );
+    event LogSwapoutFail(
+        address indexed token,
+        address indexed sender,
+        address indexed receiver,
+        uint256 amount,
+        uint256 toChainId,
+        int256[] weights
+    );
+
+    modifier onlyDistributor() {
+        require(msg.sender == distributor, "AnycallClient: onlyDistributor");
+        _;
+    }
+
+    constructor(
+        address _admin,
+        address _callProxy,
+        address _distributor
+    ) AnycallClientBase(_admin, _callProxy) {
+        distributor = _distributor;
+    }
+
+    function setTokenPeers(
+        address srcToken,
+        uint256[] calldata chainIds,
+        address[] calldata dstTokens
+    ) external onlyAdmin {
+        require(chainIds.length == dstTokens.length);
+        for (uint256 i = 0; i < chainIds.length; i++) {
+            tokenPeers[srcToken][chainIds[i]] = dstTokens[i];
+        }
+    }
+
+    function setDistributor(address _distributor) external onlyAdmin {
+        require(_distributor != address(0));
+        distributor = _distributor;
+    }
+
+    /// @dev Call by the user to submit a request for a cross chain interaction
+    function bridge(
+        address token,
+        uint256 amount,
+        address receiver,
+        uint256 toChainId,
+        int256[] memory weights,
+        uint256 periodId
+    ) external payable //onlyDistributor whenNotPaused(PAUSE_SWAPOUT_ROLE) 
+    {
+        address clientPeer = clientPeers[toChainId];
+        require(clientPeer != address(0), "AnycallClient: no dest client");
+
+        address dstToken = tokenPeers[token][toChainId];
+        require(dstToken != address(0), "AnycallClient: no dest token");
+
+        uint256 oldCoinBalance;
+        if (msg.value > 0) {
+            oldCoinBalance = address(this).balance - msg.value;
+        }
+
+        address _underlying = _getUnderlying(token);
+        if (
+            _underlying != address(0) &&
+            IERC20(token).balanceOf(msg.sender) < amount
+        ) {
+            uint256 old_balance = IERC20(_underlying).balanceOf(token);
+            IERC20(_underlying).safeTransferFrom(msg.sender, token, amount); 
+            uint256 new_balance = IERC20(_underlying).balanceOf(token);
+
+            // update amount to real balance increasement (some token may deduct fees)
+            amount = new_balance > old_balance ? new_balance - old_balance : 0;
+        } else {
+            assert(IAnyswapToken(token).burn(msg.sender, amount));
+        }
+
+        bytes memory data = abi.encodeWithSelector(
```

Comment:
Perhaps it would be helpful to include a comment on the structure of the `data` object (interfaces, etc)